src/client/chat_box/chat_box_view.tsx

```diff
@@ -0,0 +1,78 @@
+import React, {useEffect, useState} from 'react';
+import {ChatClient} from '_client/chat_box/chat_client';
+import {ChatMessage} from '_common/chat/chat_model';
+
+const INPUT_HINT = 'Type !help for commands';
+
+const messageContainerStyle: React.CSSProperties = {
+  overflowY: 'auto',
+  height: '100%',
+};
+
+export interface ChatBoxViewProps {
+  visible: boolean;
+  chatClient: Promise<ChatClient>;
+}
+
+export function ChatBoxView(props: ChatBoxViewProps): JSX.Element | null {
+  const [messages, setMessages] = useState<Array<ChatMessage>>([]);
+  const [inputText, setInputText] = useState('');
+
+  useEffect(() => {
+    const subscription = (message: ChatMessage) =>
+      setMessages([message].concat(messages));
+    props.chatClient.then((client) => client.getMessageUpdates(subscription));
```

Comment:
@ReidAnderson Do you know of a better way to handle this (lines 24-27)? 

Basically, I need to subscribe to this socket, and the socket needs to update the `messages` state. However, now this means that my `useEffect` runs every time the `chatClient` changes OR every time `messages` changes. So this means I'm removing listeners and re-attaching them after every new message.

This isn't a huge deal, but I could see some possible race conditions where we lose a message.

Reply:
Note to self - we also probably have a race condition where on a state update, we could run the cleanup, scheduling (but not actually running, yet) the unsubscribe. And then we enter the useEffect for the new update, where we subscribe, and only then do we unsubscribe. 

We could resolve this by only removing a *particular* listener, or passing chatClient in only once the Promise has been resolved. We could also probably mitigate this issue if we manage to set it up so the sub / unsub only happens once the component is first rendered, and finally removed.

Reply:
I'm not sure if this is better practice or not - but I'm pretty sure that you can define multiple `useEffect`s within a single function. So you can set up your `chatClient` code to only run on component mount/unmount (just pass `[]` as the second parameter), while you keep your message effect running every time the array changes.

Reply:
Final example in this section I believe does roughly what we're looking for here: https://reactjs.org/docs/hooks-faq.html#what-can-i-do-if-my-effect-dependencies-change-too-often